src/connection/host_lookup.c

```diff
@@ -0,0 +1,411 @@
+  hp.h_aliases = (char **) malloc (sizeof (char *));
+  hp.h_aliases[0] = (char *) malloc (sizeof (char) * HOSTNAME_BUF_SIZE);
+  hp.h_addr_list = (char **) malloc (sizeof (char *) * HOSTNAME_BUF_SIZE);
+  hp.h_addr_list[0] = (char *) malloc (sizeof (char) * IPADDR_LEN);
+
+  strcpy (hp.h_name, hostent_List[hostent_list_index].hostname);
+  strcpy (hp.h_aliases[0], hostent_List[hostent_list_index].hostname);
+  memcpy (hp.h_addr_list[0], addr_transform_buf2, sizeof (hp.h_addr_list[0]));
+
+  return &hp;
+}
+
+static int
+host_conf_load ()		//set hash table to discover error
+{
+  FILE *fp;
+  char file_line[HOSTNAME_BUF_SIZE + 1];
+  char line_buf[HOSTNAME_BUF_SIZE + 1];
+  char host_conf_file_full_path[PATH_MAX];
+  char *hosts_conf_dir;
+
+  char *token;
+  char *save_ptr_strtok;
+  /*delimiter */
+  char *delim = " \t\n";
+  int host_count = 0;
+
+  /*False, if hostent_List[index].hostname is set */
+  bool storage_flag = 0;
+
+  memset (file_line, 0, HOSTNAME_BUF_SIZE + 1);
+  memset (line_buf, 0, HOSTNAME_BUF_SIZE + 1);
+
+  host_count = host_conf_element_Count;
+  assert (!host_count);
+
+  hosts_conf_dir = envvar_confdir_file (host_conf_file_full_path, PATH_MAX, "hosts.conf");
+  fp = fopen (hosts_conf_dir, "r");
+  if (fp == NULL)
+    {
+      return -1;
+      //error msg-> temp
+    }
+
+  while (fgets (file_line, HOSTNAME_BUF_SIZE, fp) != NULL)
+    {
+//verifying duplicated hostname will be implemented soon
+      if (file_line[0] == '#')
+	continue;
+      if (file_line[0] == '\n')
+	continue;
+
+      token = strtok_r (file_line, delim, &save_ptr_strtok);
+      storage_flag = false;
+
+      do
+	{
+	  if (*token == '\0' || *token == '#')
+	    {
+	      break;
+	    }
+	  if (!storage_flag)
+	    {
+	      strcpy (hostent_List[host_count].addr, token);
+	      storage_flag = true;
+	    }
+	  else
+	    {
+	      strcpy (hostent_List[host_count].hostname, token);
+	    }
+	}
+      while (token = strtok_r (NULL, delim, &save_ptr_strtok));
+//naming rule check, detailed rule, error set
+//if hostent_List[host_count].hostname is empty, set error
+      host_count++;
+    }
+  return host_count;
+}
+
+/*
+ * gethostbyname_uhost () - Do same job with gethostbyname (), using by the 'user' defined 'hosts.conf' file.
+ * 
+ * return   : the hostent pointer.
+ * hostname (in) : the hostname.
+ *
+ */
+hostent *
+gethostbyname_uhost (char *hostname)
+{
+  hostent *hp;
+
+  if (host_conf_Use < 0)
+    {
+      host_conf_Use = prm_get_bool_value (PRM_ID_USE_USER_HOSTS);
+    }
+
+  if (host_conf_Use == 0)
+    {
+      return gethostbyname (hostname);
+    }
+
+  if (!host_conf_element_Count)
+    {
+      host_conf_element_Count = host_conf_load ();
+      if (host_conf_element_Count <= 0)
+	{
+	  return NULL;
+	}
+    }
+
+  hp = host_lookup_internal (hostname, NULL, HOSTNAME_LOOKUP);
+
+  if (hp == NULL)
+    {
+      return NULL;
+    }
+
+  return hp;
+}
+
+int
+gethostbyname_r_uhost (const char *hostname, struct hostent *out_hp)
```

Comment:
It is necessary to define gethost_r for each GLIBC, SOLARIS, and HOSTENT_DATA. This is because each argument has a different number and form. For example,

```
#ifdef HAVE_GETHOSTBYNAME_R
#if defined (HAVE_GETHOSTBYNAME_R_GLIBC)
int
gethostbyname_r(const char *name,
               struct hostent *ret, char *buf, size_t buflen,
               struct hostent **result, int *h_errnop)
#elif defined (HAVE_GETHOSTBYNAME_R_SOLARIS)
gethostbyname_r(const char *name,
               struct hostent *ret, char *buf, size_t buflen,
               int *h_errnop)
#elif defined (HAVE_GETHOSTBYNAME_R_HOSTENT_DATA)
gethostbyname_r(const char *name, struct hostent *ret, struct hostent_data *ht_data)
#else
#error "HAVE_GETHOSTBYNAME_R"
#endif /* HAVE_GETHOSTBYNAME_R_GLIBC */
{
	/* cpdes for */
}
#endif /* HAVE_GETHOSTBYNAME_R */
```

Reply:
I applied it. Thank you!